src/lua/uri.c

```diff
@@ -0,0 +1,350 @@
+	if (uristr != NULL) {
+		rc = uri_create(uri, uristr);
+		if (rc != 0) {
+			diag_set(IllegalParams, "Incorrect URI: expected "
+				 "host:service or /unix.socket");
+		}
+	} else {
+		diag_set(IllegalParams, "Incorrect type for URI in nested "
+			 "table: should be string, number");
+		rc = -1;
+	}
+	lua_pop(L, 1);
+	if (rc != 0)
+		return rc;
+	lua_pushstring(L, "params");
+	lua_rawget(L, idx);
+	rc = uri_add_params_from_lua(uri, L, true);
+	lua_pop(L, 1);
+	if (rc != 0)
+		uri_destroy(uri);
+	return rc;
+
+}
+
+/**
+ * Create @a uri from the value at the given valid @a idx, which
+ * should be positive value.
+ */
+static int
+luaT_uri_create(struct lua_State *L, int idx, struct uri *uri)
+{
+	int rc = 0;
+	if (lua_isstring(L, idx)) {
+		rc = uri_create(uri, lua_tostring(L, idx));
+		if (rc != 0) {
+			diag_set(IllegalParams, "Incorrect URI: "
+				 "expected host:service or "
+				 "/unix.socket");
+		}
+	} else if (lua_istable(L, idx)) {
+		rc = uri_create_from_lua_table(L, idx, uri);
+	} else if (lua_isnil(L, idx)) {
+		uri_create(uri, NULL);
+	} else {
+		diag_set(IllegalParams, "Incorrect type for URI: "
+			 "should be string, number or table");
+		rc = -1;
+	}
+	assert(rc == 0 || !diag_is_empty(diag_get()));
+	return rc;
+}
+
+/**
+ * Create @a uri_set from the table, which located at the given valid @a idx,
+ * which should be positive value.
+ */
+static int
+uri_set_create_from_lua_table(struct lua_State *L, int idx,
+			      struct uri_set *uri_set)
+{
+	int rc = 0;
+	assert(lua_type(L, idx) == LUA_TTABLE);
+	int size = lua_objlen(L, idx);
+	struct uri uri;
+
+	uri_set_create(uri_set, NULL);
+	if (is_field_present(L, idx, "uri") ||
+	    (size == 1 && field_type(L, idx, 1) != LUA_TTABLE)) {
+		rc = luaT_uri_create(L, idx, &uri);
+		if (rc == 0)
+			uri_set_add(uri_set, &uri);
+		return rc;
+	} else if (size == 0) {
+		return 0;
+	}
+
+	/*
+	 * All numeric keys corresponds to URIs in string or table
+	 * format.
+	 */
+	for (int i = 0; i < size && rc == 0; i++) {
+		lua_rawgeti(L, idx, i + 1);
+		rc = luaT_uri_create(L, lua_gettop(L), &uri);
+		if (rc == 0)
+			uri_set_add(uri_set, &uri);
+		lua_pop(L, 1);
+	}
+	if (rc != 0)
+		goto fail;
+
+	/*
+	 * Here we are only in case when it is an URI array, so it
+	 * shouldn't be "params" field here.
+	 */
+	if (is_field_present(L, idx, "params")) {
+		diag_set(IllegalParams, "URI query parameters are "
+			 "not allowed for multiple URIs");
+		goto fail;
+	}
+
+	lua_pushstring(L, "default_params");
+	lua_rawget(L, idx);
+	if (!lua_isnil(L, -1)) {
+		for (int i = 0; i < uri_set->uri_count && rc == 0; i++) {
+			struct uri *uri = &uri_set->uris[i];
+			rc = uri_add_params_from_lua(uri, L, false);
+			assert(rc == 0 || !diag_is_empty(diag_get()));
+		}
+	}
+	lua_pop(L, 1);
+	if (rc != 0)
+		goto fail;
+	return 0;
+fail:
+	uri_set_destroy(uri_set);
+	return -1;
+}
+
+/**
+ * Create @a uri_set from the value at the given valid @a idx, which should be
+ * positive value.
```

Comment:
If this function doesn't work if idx is negative, you should add an assertion. Anyway, I don't understand why this function wouldn't work in this case. If some of the functions it calls needs a positive idx, you can make it positive there using `lua_gettop`.

Reply:
Yes you right, i understand my mistake and rework

Reply:
Scattering idx checks all over the code is fragile and inconvenient - please instead add something like this to the beginning of each function that expects a positive index:

```C
if (idx < 0)
    idx = lua_gettop(L) + idx + 1;
assert(idx > 0);
```

There's the only such function AFAICS - `luaT_uri_set_create`. All other functions can use `assert(idx > 0);`, because they are for internal use only.

Reply:
fixed, also this necessary for luaT_uri_create